src/rust/engine/process_execution/src/remote.rs

```diff
@@ -1315,6 +1236,94 @@ pub fn extract_output_files(
   .to_boxed()
 }
 
+/// Check the remote Action Cache for a cached result of running the given `action_digest`.
+///
+/// This check is necessary because some RE servers do not short-circuit the Execute method
+/// by checking the Action Cache (e.g., BuildBarn). Thus, this client must check the cache
+/// explicitly in order to avoid duplicating already-cached work. This behavior matches
+/// the Bazel RE client.
+pub async fn check_action_cache(
+  action_digest: Digest,
+  metadata: &ProcessMetadata,
+  platform: Platform,
+  context: &Context,
+  action_cache_client: Arc<ActionCacheClient>,
+  headers: &BTreeMap<String, String>,
+  store: Store,
+) -> Result<Option<FallibleProcessResultWithPlatform>, String> {
+  let mut request = bazel_protos::remote_execution::GetActionResultRequest::new();
+  if let Some(ref instance_name) = metadata.instance_name {
+    request.set_instance_name(instance_name.clone());
+  }
+  request.set_action_digest(action_digest.into());
+
+  let call_opt = call_option(headers, Some(context.build_id.clone()))?;
+
+  let action_result_response = action_cache_client
+    .get_action_result_async_opt(&request, call_opt)
+    .unwrap()
+    .compat()
+    .await;
+
+  match action_result_response {
+    Ok(action_result) => {
+      let response =
+        populate_fallible_execution_result(store.clone(), &action_result, vec![], platform, false)
+          .compat()
+          .await?;
+      Ok(Some(response))
+    }
+    Err(err) => match err {
+      grpcio::Error::RpcFailure(rpc_status)
+        if rpc_status.status == grpcio::RpcStatusCode::NOT_FOUND =>
+      {
```

Comment:
nit - it's fine to leave off the brackets here and just return `Ok(None)` directly from this match arm